Multimonth calendar view goodie

rubinovitz opened the pull request:
cc @russellholt 
What does your Instant Answer do?

Show a year view of calendars

What problem does your Instant Answer solve (Why is it better than organic links)?

Full calendar views for a year are hard to find

What is the data source for your Instant Answer? (Provide a link if possible)

Perl Datetime and Javascript moment.js

Are there any other alternative (better) data sources?

It would be cool to add holidays at some point. This data is pretty standard.

What are some example queries that trigger this Instant Answer?
[cal | calendar] month year
[cal | calendar]
[cal | calendar] month
[cal | calendar] year

Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)

Everyone who wants to keep track of time.

Is this Instant Answer connected to a DuckDuckHack Instant Answer idea?

Yes it redoes the current calendar. (Speaking of which we may want to rename this to Calendar)

Which existing Instant Answers will this one supercede/overlap with?

CalendarToday

Are you having any problems? Do you need our help with anything?
-How do I best test a primarily javascript based goodie?
-I need help testing the async moment injection since it's not available externally yet
- Should I still be generating plain text?
- Is the selected day working? I can't see on my end.

What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)
![screenshot 2015-08-19 13 43 37](https://cloud.githubusercontent.com/assets/1807895/9364378/076a4d4e-467a-11e5-8f8c-0f68159d1141.png)
![screenshot 2015-08-19 13 43 24](https://cloud.githubusercontent.com/assets/1807895/9364379/076bdc4a-467a-11e5-88f6-0124518aadf0.png)
![screenshot 2015-08-19 13 42 55](https://cloud.githubusercontent.com/assets/1807895/9364380/076d2f96-467a-11e5-857f-117f6cde38c5.png)

Checklist

Please place an 'X' where appropriate.

- [X] Added metadata and attribution information (https://duck.co/duckduckhack/metadata)
- [] Wrote test file and added to t/ directory (https://duck.co/duckduckhack/goodie_tests)
Unsure what to test here
- [X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
- [X] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
- [X] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        - [] Google Chrome
        - [] Firefox
        - [] Opera
        - [] IE 10

    - Windows 7
        - [] Google Chrome
        - [] Firefox
        - [] Opera
        - [] IE 8
        - [] IE 9
        - [] IE 10

    - Windows XP
        - [] IE 7
        - [] IE 8

    - Mac OSX
        - [X] Google Chrome
        - [X] Firefox
        - [] Opera
        - [X] Safari

    - iOS (iPhone)
        - [] Safari Mobile
        - [] Google Chrome
        - [] Opera

    - iOS (iPad)
        - [] Safari Mobile
        - [] Google Chrome
        - [] Opera

    - Android
        - [] Firefox
        - [] Native Browser
        - [] Google Chrome
        - [] Opera

https://duck.co/ia/view/calendar_today

moollaza commented:
@rubinovitz thanks a lot! We'll take a look at this soon and give some feedback. 

BTW I noticed the tests are failing, can you update them so they're passing? You can run them locally with `duckpan test` or `prove -Ilib t/Calendar.t`

rubinovitz commented:
@moollaza  one of my questions in the PR is asking how i should be testing this

russellholt commented on share/goodie/calendar_today/calendar_today.js:
```diff
@@ -0,0 +1,163 @@
+            id: "calendar",
+            name: "Calendar",
+            meta:{
+                idField: currentMonth,
```
needs to be `idField: 'id'`

russellholt commented on share/goodie/calendar_today/calendar_today.js:
```diff
@@ -0,0 +1,163 @@
+            startDate = startDate.add(1, 'months');
+        }
+
+        var f = DDH_async_add || Spice.add;        
```
remove this for now, until the async thing is merged.

russellholt commented on share/goodie/calendar_today/calendar_today.js:
```diff
@@ -0,0 +1,163 @@
+DDH.calendar_today = DDH.calendar_today || {}; // create the namespace in case it doesn't exist
+
+DDH.calendar_today.build = DDH.calendar_today.build_async =  function(ops, DDH_async_add) {
```
comment out the build_async until it's ready.

moollaza commented on lib/DDG/Goodie/CalendarToday.pm:
```diff
@@ -5,8 +5,9 @@ use strict;
 use Try::Tiny;
-use URI::Escape::XS qw(encodeURIComponent);
 use Text::Trim;
+use URI::Escape::XS qw(encodeURIComponent);
```
We can remove this now. We're no longer encoding any HTML (thankfully!)

bsstoner commented on share/goodie/calendar_today/calendar_today.css:
```diff
@@ -1,20 +1,70 @@
+    line-height: 2.17;
+    weight: 400;
+    padding: 1.25em;
 }
```
Should this be namespaced? `.zci--calendar` or something?

mintsoft replied:
Yes it should be `.zci--calendar .calendar`

rubinovitz replied:
This has been added. Thanks.

bsstoner commented on share/goodie/calendar_today/calendar_today.js:
```diff
@@ -0,0 +1,162 @@
+            item = {}; // to hold a month at a time
+           
+        // get january of the given year
+        var startDate = moment([currentYear, 0]);
```
Does this mean we'd only show the current calendar year? That might not be ideal in December where you'd want to see first few months of the next year.

russellholt replied:
@bsstoner  can the scrolling controls be used to add more data?

bsstoner replied:
hmm... we don't have a good API for that right now. That's a good idea though.

rubinovitz replied:
@bsstoner yes we start at Jan of the year by default.

rubinovitz replied:
@bsstoner looking back at the original idea our intention was to also have all calendar queries producing a tile view of 12 months, with a dropdown to pick a different year. However, the dropdown did not get implemented in this iteration. Do you have any pointers to documentation as to how such a dropdown could be feasible? I think it's a potential solution for this problem (you chose December, don't see the months you wanted in this view, and go to the next year).

mintsoft commented on lib/DDG/Goodie/CalendarToday.pm:
```diff
@@ -45,85 +46,43 @@ handle remainder => sub {
+            my ($currentDay, $currentMonth, $currentYear) = ($date_object->day(), $date_object->month(), $date_object->year());
+        } else {
+            # check if 4 digit year is being queried e.g. cal 1999
+            if ($query =~ qr/\d\d\d\d/){
```
You could use `qr/\d{4}/`

mintsoft commented on lib/DDG/Goodie/CalendarToday.pm:
```diff
@@ -45,85 +46,43 @@ handle remainder => sub {
-              )->day(),
-            highlight => $highlightDay,
-        });
+                day => $currentDay,
```
If the query is in the `\d{4}` format, it looks like `$currentDay` is unset?

rubinovitz replied:
Good point. I suppose if the current year is specified currentDay should be set. I'll get on that.

rubinovitz replied:
@mintsoft I actually think this might be irrelevant as I've been thinking of this as we weren't thinking about day specific queries, just highlighting the "current day" if it is displayed. Since this is the case, I've been getting the current day and highlighting days that match that on the front end.

@russellholt Can you please confirm this works or should we be able to pass in a specific day other than 'today' to highlight from the backend?

russellholt replied:
I'm sure only the current day matters - no need to query a soecific day for now. open to other ideas but I'd push them to a future version.

mintsoft commented on share/goodie/calendar_today/calendar_today.css:
```diff
@@ -1,20 +1,70 @@
-    font-size: 1.25em;
-}
 
+/*
```
I'd probably bomb the commented bit completely

mintsoft commented on share/goodie/calendar_today/calendar_today.js:
```diff
@@ -0,0 +1,162 @@
+            var daysArray = [];
+            //get start and end of this month and last month
+            var endDate = moment(startDate).endOf('month'),
+                startOfLastMonth = startDate.clone().subtract(1, 'months').startOf('month'),
```
Is this definitely 1 calendar month? i.e. 31st march - 1 month = 28th feb?

rubinovitz replied:
@mintsoft Yes, moment.js accounts for this. They go into it a bit in the documentation under [add](http://momentjs.com/docs/#/manipulating/add/) since its used for subtracting.

Also example in console:

![screenshot 2015-08-24 23 52 30](https://cloud.githubusercontent.com/assets/1807895/9458337/347fbacc-4abb-11e5-86e3-5245221314e7.png)

mintsoft commented:
@rubinovitz It looks like the tests haven't been updated: https://travis-ci.org/duckduckgo/zeroclickinfo-goodies/builds/76332952#L4028

moollaza commented:
>@moollaza one of my questions in the PR is asking how i should be testing this

Whoops, sorry @rubinovitz I didn't see that!

Goodies can only verify that the object `return`'ed from the `handle` function matches what the test expects. There's currently no way to test the frontend JS output or display (we're working on it though!).

You can model your test after something like `Minecraft.t.` I fiddled around with the test file and arrived at this which should be a good starting point: https://gist.github.com/moollaza/44f63947171bcacc3e23

moollaza commented on lib/DDG/Goodie/CalendarToday.pm:
```diff
@@ -45,85 +46,43 @@ handle remainder => sub {
-    my $next = $firstDay->clone->add(months => 1);
-
-    # Print heading
-    my $rText = "\n";
```
@russellholt do we want to preserve the plain text result we are currently providing to /lite and /html?

Right now we'll be returning the string "Calendar Today" to /lite and /html (see https://github.com/duckduckgo/zeroclickinfo-goodies/pull/1417/files#diff-34f2e3cbdc7d6f82d8f17af4a4a16eabR68)

russellholt replied:
Losing lite/html is a drawback of the JS approach, as we've discussed many times. I think we're still on the fence about it, but leaning towards removing it. We could easily add the previous behavior back, just for the old world, but I don't see a huge value in it for this IA in particular. Open to contrary opinions.

moollaza commented on lib/DDG/Goodie/CalendarToday.pm:
```diff
@@ -45,85 +46,43 @@ handle remainder => sub {
+       return $plaintext,
+        structured_answer => {
+            id => 'calendar_today',
+            name => 'Answer',
```
This should be removed so it doesn't conflict with your `name: "Calendar"` in the JS

rubinovitz replied:
can you please be more specific as to what i should remove? I get throw an error when removing name and id.

```
Use of uninitialized value in concatenation (.) or string at /Users/j/perl5/lib/perl5/App/DuckPAN/Web.pm line 426

```

russellholt replied:
this one is fine, what has to match is this with the one in the argument to the build_async call, https://github.com/duckduckgo/zeroclickinfo-goodies/pull/1417/files#diff-77ab5e499ea538b95ba9bb3084e64216R141

and that has to match the one on the IA Page, https://duck.co/ia/view/calendar_today

russellholt commented on share/goodie/calendar_today/calendar_today.js:
```diff
@@ -0,0 +1,162 @@
+         * @param {moment object} day
+         * @return {Object} day, classes, date
+         */
+        createDayObject = function(day) {
```
var

russellholt commented on share/goodie/calendar_today/calendar_today.js:
```diff
@@ -0,0 +1,162 @@
+                endOfNextMonth = startOfNextMonth.clone().endOf('month');
+
+            // get start date to iterate from
+            currentIntervalStart = startDate.clone();
```
this is a global

russellholt commented on share/goodie/calendar_today/calendar_today.js:
```diff
@@ -0,0 +1,162 @@
+            var extraClasses = '';
+
+            // check if day is not from this month
+            if (currentIntervalStart.month() !== day.month()){
```
references the global defined below

rubinovitz replied:
Will move to argument.

russellholt commented on share/goodie/calendar_today/calendar_today.js:
```diff
@@ -0,0 +1,162 @@
+            item.daysOfTheWeek = daysOfTheWeek;
+            item.year = year;
+            // add 1 since perl does. this is used for selected item
+            item.id = month + 1;
```
:+1: would add a comment about why the `id` exists - for selectedItem in the tile view

russellholt commented on t/CalendarToday.t:
```diff
@@ -9,106 +9,35 @@ use Test::MockTime qw( :all );
+
+    return structured_answer => {
+        id => 'calendar_today',
+        name => 'Answer',
```
can remove `name`

russellholt commented on t/CalendarToday.t:
```diff
@@ -9,106 +9,35 @@ use Test::MockTime qw( :all );
+            month => $month,
+            year => $year
+        },
+        templates => {
```
`templates => 1` here while the templates are defined in the corresponding JS

russellholt commented on share/goodie/calendar_today/calendar_today.js:
```diff
@@ -0,0 +1,161 @@
+        // add spice
+       DDH_async_add({
+            id: "calendar",
+            name: "Calendar",
```
can remove `name` while the tab is define on the IA page.

russellholt commented on share/goodie/calendar_today/calendar_today.js:
```diff
@@ -0,0 +1,161 @@
+        }
+        // add spice
+       DDH_async_add({
+            id: "calendar",
```
needs to match the IA id, which is `calendar_today`

russellholt commented:
Hey looks like the current month is still selected when another year is given

rubinovitz commented:
@russellholt yes while making that change and not always setting item ids I realized the build_async was firing twice and the calendar would render twice when I did not set ids, so I didn't end up pushing it up.

russellholt commented:
I don't see this firing twice behavior you're describing for build_async

rubinovitz commented:
@russellholt I updated my duckPAN, removed item.id (which was causing it to only show uniquely id-ed tiles) and added logging inside of build_async that is logging twice for me. Is it still not firing twice on your end?

jagtalon commented:
Hey, @rubinovitz! Sorry for chiming in late. 

- I'm not quite sure if I understand why `item.id = month + 1` was removed. It looks like there's an error when there's no `id` and there are more than 12 tiles:

Before: :anguished:

![screen shot 2015-09-08 at 9 41 00 pm](https://cloud.githubusercontent.com/assets/81969/9751423/95a4a03e-5672-11e5-8690-4e9349b5f8db.png)

After: :smirk:

![screen shot 2015-09-08 at 9 40 15 pm](https://cloud.githubusercontent.com/assets/81969/9751424/95a56794-5672-11e5-8136-2c8ad0e1f6b4.png)

jagtalon commented:
@rubinovitz Looks like a bunch of numbers disappear when the theme changes:

![screen shot 2015-09-08 at 9 50 53 pm](https://cloud.githubusercontent.com/assets/81969/9751537/0b7d29b0-5674-11e5-8c36-6b49a4628143.png)

This is easy to fix, though. Check out the "Text Colors" section in https://duckduckgo.com/styleguide.

![screen shot 2015-09-08 at 9 52 51 pm](https://cloud.githubusercontent.com/assets/81969/9751541/1f7cb91c-5674-11e5-8227-e11d21efff50.png)

Use these classes in your HTML because they're magical :sparkler:. They change when the theme changes.

jagtalon commented on share/goodie/calendar_today/calendar_today.js:
```diff
@@ -0,0 +1,161 @@
+            // push item into items
+            items.push(item);
+            // clear item
+            item = {};
```
Nitpick: Is it possible to do something like this instead? Looks cleaner that way. :)

```javascript
items.push({
    id: month + 1,
    month_name: month_name,
    ...
});
```

jagtalon commented on share/goodie/calendar_today/calendar_today.css:
```diff
@@ -1,53 +1,63 @@
+    text-align: center; 
+    color: #FFF;
+    border-radius: 50%;
+    -webkit-border-radius: 50%;
```
No need for these two since `.circle` already adds the `border-radius`

jagtalon commented:
Closing this. If anyone wants to continue finishing this the code is on https://github.com/duckduckgo/zeroclickinfo-goodies/tree/ideas/multimonth-cal!

rubinovitz commented:
@jagtalon 
Re "I'm not quite sure if I understand why item.id = month + 1 was removed. It looks like there's an error when there's no id and there are more than 12 tiles:"

That was to show that the build_async has an error where it runs twice (you can see it in your screenshots in the console logs). Just an FYI if anyone ends up finishing this, that needs to be fixed internally first.

jagtalon commented:
Hey, @rubinovitz! Sorry for not being clear--I meant this line https://github.com/rubinovitz/zeroclickinfo-goodies/blob/ad7f7c9fb6c14fb1714f560fa5c67251dd50b430/share/goodie/calendar_today/calendar_today.js#L131.

Would you be interested in finishing this? :)

rubinovitz commented:
I uncommented that line to try to make it more apparent that build_async
was running twice after trying to get it fixed internally multiple times.

No thanks.

On Fri, Sep 18, 2015, 4:42 PM Jag Talon <notifications@github.com> wrote:

> Hey, @rubinovitz <https://github.com/rubinovitz>! Sorry for not being
> clear--I meant this line
> https://github.com/rubinovitz/zeroclickinfo-goodies/blob/ad7f7c9fb6c14fb1714f560fa5c67251dd50b430/share/goodie/calendar_today/calendar_today.js#L131
> .
>
> Would you be interested in finishing this? :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/duckduckgo/zeroclickinfo-goodies/pull/1417#issuecomment-141563087>
> .
>

rubinovitz commented:
s/uncommented/removed

jagtalon commented:
@rubinovitz ok thank you!
